overlord/hookstate/hooks.go

```diff
@@ -72,6 +78,189 @@ func SetupPreRefreshHook(st *state.State, snapName string) *state.Task {
+	switch action {
+	case snapstate.GateAutoRefreshHold:
+		// for action=hold the ctlcmd calls HoldRefresh; only unlock runinhibit.
+		if h.refreshAppAwareness {
+			if err := runinhibit.Unlock(snapName); err != nil {
+				return fmt.Errorf("cannot unlock inhibit lock of snap %s: %v", snapName, err)
+			}
+		}
+	case snapstate.GateAutoRefreshProceed:
+		// for action=proceed the ctlcmd doesn't call ProceedWithRefresh
+		// immediately, do it here.
+		if err := snapstate.ProceedWithRefresh(st, snapName); err != nil {
+			return err
+		}
+		if h.refreshAppAwareness {
+			// we have HintInhibitedGateRefresh lock already when running the hook,
+			// change it to HintInhibitedForRefresh.
+			if err := runinhibit.LockWithHint(snapName, runinhibit.HintInhibitedForRefresh); err != nil {
+				return fmt.Errorf("cannot set inhibit lock for snap %s: %v", snapName, err)
+			}
+		}
+	default:
+		return fmt.Errorf("internal error: unexpected action %v", action)
+	}
+
+	return nil
+}
+
+// Error handles gate-auto-refresh hook failure; it assumes hold.
+func (h *gateAutoRefreshHookHandler) Error(hookErr error) (err error) {
+	ctx := h.context
+	st := h.context.State()
+	ctx.Lock()
+	defer ctx.Unlock()
+
+	snapName := h.context.InstanceName()
+
+	var lock *osutil.FileLock
+
+	// the refresh is going to be held, release runinhibit lock.
+	if h.refreshAppAwareness {
+		// obtain snap lock before manipulating runinhibit lock.
+		lock, err = snaplock.OpenLock(snapName)
+		if err != nil {
+			return err
+		}
+		if err := lock.Lock(); err != nil {
+			return err
+		}
+		defer lock.Unlock()
+
+		if err := runinhibit.Unlock(snapName); err != nil {
+			return fmt.Errorf("cannot release inhibit lock of snap %s: %v", snapName, err)
+		}
+	}
+
+	if a := ctx.Cached("action"); a != nil {
+		action, ok := a.(snapstate.GateAutoRefreshAction)
+		if !ok {
+			return fmt.Errorf("internal error: unexpected action type %T", a)
+		}
+		// nothing to do if the hook already requested hold.
+		if action == snapstate.GateAutoRefreshHold {
+			return nil
+		}
+	}
+
+	// the hook didn't request --hold, or it was --proceed. since the hook
+	// errored out, assume hold.
+
+	var affecting []string
+	if err := ctx.Get("affecting-snaps", &affecting); err != nil {
+		return fmt.Errorf("internal error: cannot get affecting-snaps")
+	}
+
+	// no duration specified, use maximum allowed for this gating snap.
+	var holdDuration time.Duration
+	if err := snapstate.HoldRefresh(st, snapName, holdDuration, affecting...); err != nil {
+		// note, previous hook error (hookErr) is going to be logged by hookmgr
+		// after this Error() handler, so only log the new error.
+		h.context.Errorf("error: %v (while handling previous hook error)", err)
+		if _, ok := err.(*snapstate.HoldError); ok {
+			return nil
+		}
+		// anything other than HoldError becomes an error of the handler.
+		return err
+	}
+
+	// TODO: consider assigning a special health state for the snap.
+
+	return nil
```

Comment:
we forgot to consider something,  the error will still make the whole autorefresh fail, no?  it seems like we need  way to tell HookManager to ignore the error like IgnoreError but going through the handlers anyway?

Reply:
Updated.